### PR TITLE
Added pagination for Db2 connector

### PR DESCRIPTION
**Pull request by VenkatasivareddyTR**
*Issue #, if available:*
DB2 Connector pagination is not implemented.

*Description of changes:*
Added pagination for Db2 connector

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java [hunk 1]
```diff
@@ -350,5 +350,56 @@ public void doListTables() throws Exception {
...
+        resultSet = mockResultSet(schema, values, new AtomicInteger(-1));
+        Mockito.when(preparedStatement.executeQuery()).thenReturn(resultSet);
+        listTablesResponse = this.db2MetadataHandler.doListTables(
+                blockAllocator, new ListTablesRequest(this.federatedIdentity, "testQueryId",
+                        "testCatalog", "testSchema", "1", 1));
+        Assert.assertEquals("2", listTablesResponse.getNextToken());
+        Assert.assertArrayEquals(expected, listTablesResponse.getTables().toArray());
+
+
+        // Test 3: Testing single table returned when requesting pageSize 2 signifying end of pagination where nextToken is null.
```
Comment (samarsar) [inline athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java:390]: Can you add a test case for when pageSize is UNLIMITED but the nextToken begins at something other than null/0?

**Review by samarsar — CHANGES_REQUESTED** [review]

File: athena-db2/src/main/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandler.java [hunk 2]
```diff
@@ -163,12 +165,50 @@ public ListTablesResponse doListTables(final BlockAllocator blockAllocator, fina
     {
         try (Connection connection = getJdbcConnectionFactory().getConnection(getCredentialProvider())) {
             LOGGER.info("{}: List table names for Catalog {}, Schema {}", listTablesRequest.getQueryId(), listTablesRequest.getCatalogName(), listTablesRequest.getSchemaName());
-            List<String> tableNames = getTableList(connection, Db2Constants.QRY_TO_LIST_TABLES_AND_VIEWS, listTablesRequest.getSchemaName());
-            List<TableName> tables = tableNames.stream().map(tableName -> new TableName(listTablesRequest.getSchemaName(), tableName)).collect(Collectors.toList());
-            return new ListTablesResponse(listTablesRequest.getCatalogName(), tables, null);
+
+            int pageSize = listTablesRequest.getPageSize();
+            if (pageSize == UNLIMITED_PAGE_SIZE_VALUE || pageSize == 0) {
```
[the remark this answers is not in the document]
Comment (VenkatasivareddyTR) [reply athena-db2/src/main/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandler.java]: Thank you @samarsar addressed your review comments.

File: athena-db2/src/main/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandler.java [hunk 3]
```diff
@@ -163,12 +165,50 @@ public ListTablesResponse doListTables(final BlockAllocator blockAllocator, fina
...
+        String token = listTablesRequest.getNextToken();
+        int pageSize = listTablesRequest.getPageSize();
+
+        int t = token != null ? Integer.parseInt(token) : 0;
+
+        LOGGER.info("Starting pagination at {} with page size {}", t, pageSize);
+        List<TableName> paginatedTables = getPaginatedTables(connection, listTablesRequest.getSchemaName(), t, pageSize);
+        LOGGER.info("{} tables returned. Next token is {}", paginatedTables.size(), t + pageSize);
+
+        String nextToken = paginatedTables.isEmpty() || paginatedTables.size() < pageSize ? null : Integer.toString(t + pageSize);
```
[the remark this answers is not in the document]
Comment (VenkatasivareddyTR) [reply athena-db2/src/main/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandler.java]: Updated code to handle this scenario.

File: athena-db2/src/main/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandler.java [hunk 4]
```diff
@@ -163,10 +165,70 @@ public ListTablesResponse doListTables(final BlockAllocator blockAllocator, fina
...
+            return listPaginatedTables(connection, listTablesRequest);
+        }
+    }
+
+    @Override
+    public ListTablesResponse listPaginatedTables(final Connection connection, final ListTablesRequest listTablesRequest) throws SQLException
+    {
+        LOGGER.debug("Starting listPaginatedTables for Db2.");
+        String token = listTablesRequest.getNextToken();
+        int pageSize = listTablesRequest.getPageSize();
```
[the remark this answers is not in the document]
Comment (VenkatasivareddyTR) [reply athena-db2/src/main/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandler.java:188]: Thank you @samarsar utilized methods from the PaginationHelper class.

**Review by samarsar — DISMISSED** [review]

File: athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java [hunk 5]
```diff
@@ -350,5 +351,92 @@ public void doListTables() throws Exception {
...
+
+        ListTablesResponse listTablesResponse = this.db2MetadataHandler.doListTables(
+                blockAllocator, new ListTablesRequest(this.federatedIdentity, "testQueryId",
+                        "testCatalog", "testSchema", null, 1));
+
+        Assert.assertEquals("1", listTablesResponse.getNextToken());
+        Assert.assertArrayEquals(expected, listTablesResponse.getTables().toArray());
+
+        // Test 2: Testing next table returned of page size 1 and nextToken 1
+        preparedStatement = Mockito.mock(PreparedStatement.class);
```
Comment (fal-bharadwaj) [inline athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java]: seems like there is a lot of duplicate code around preparedStatement mocking in each test here that can be centralized, please use a method to achieve this and pass arguments for any dynamic fields like values/expected/resultSet.

**Review by burhan94 — DISMISSED** [review]

**Comment by burhan94** [comment]
Hi, this PR has merge conflicts

**Comment by VenkatasivareddyTR** [comment]
> Hi, this PR has merge conflicts

Hi @burhan94, thank you for approving the PR. Merge conflicts have been resolved now.

**Comment by fal-bharadwaj** [comment]
Can you update the patch coverage mentioned in this CodeCov report please - https://github.com/awslabs/aws-athena-query-federation/pull/2772#issuecomment-3344837449

**Comment by VenkatasivareddyTR** [comment]
> Can you update the patch coverage mentioned in this CodeCov report please - [#2772 (comment)](https://github.com/awslabs/aws-athena-query-federation/pull/2772#issuecomment-3344837449)

Hi @fal-bharadwaj, the previous patch coverage was 86%, but it still failed. It's now updated to 100% and is passing.

File: athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java (hunk 5 above)
Comment (VenkatasivareddyTR) [reply athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java]: Thank you, @fal-bharadwaj, for the review. Created a common method for all list-paginated scenarios.

Comment (fal-bharadwaj) [reply athena-db2/src/test/java/com/amazonaws/athena/connectors/db2/Db2MetadataHandlerTest.java]: Looks good, thanks, approved!

**Review by burhan94 — APPROVED** [review]
